data/src/main/java/org/apache/iceberg/data/PartitionStatsHandler.java

```diff
@@ -135,20 +146,112 @@ public static PartitionStatisticsFile computeAndWriteStatsFile(Table table) thro
    */
   public static PartitionStatisticsFile computeAndWriteStatsFile(Table table, long snapshotId)
       throws IOException {
+    return computeAndWrite(table, snapshotId, true /* recompute */);
+  }
+
+  /**
+   * Incrementally computes the stats after the snapshot that has partition stats file till the
+   * given snapshot and writes the combined result into a {@link PartitionStatisticsFile} after
+   * merging the stats.
+   *
+   * @param table The {@link Table} for which the partition statistics is computed.
+   * @param snapshotId snapshot for which partition statistics are computed.
+   * @return {@link PartitionStatisticsFile} for the given snapshot, or null if no statistics are
+   *     present.
+   */
+  public static PartitionStatisticsFile computeAndWriteStatsFileIncremental(
+      Table table, long snapshotId) throws IOException {
+    return computeAndWrite(table, snapshotId, false /* recompute */);
+  }
+
+  private static PartitionStatisticsFile computeAndWrite(
+      Table table, long snapshotId, boolean recompute) throws IOException {
+    Preconditions.checkArgument(table != null, "Table cannot be null");
     Snapshot snapshot = table.snapshot(snapshotId);
     Preconditions.checkArgument(snapshot != null, "Snapshot not found: %s", snapshotId);
 
-    Collection<PartitionStats> stats = PartitionStatsUtil.computeStats(table, snapshot);
-    if (stats.isEmpty()) {
+    StructType partitionType = Partitioning.partitionType(table);
+    PartitionMap<PartitionStats> resultStatsMap =
+        computeStats(table, snapshot, partitionType, recompute);
+    if (resultStatsMap.isEmpty()) {
       return null;
     }
 
-    StructType partitionType = Partitioning.partitionType(table);
-    List<PartitionStats> sortedStats = PartitionStatsUtil.sortStats(stats, partitionType);
+    List<PartitionStats> sortedStats =
+        PartitionStatsUtil.sortStats(resultStatsMap.values(), partitionType);
     return writePartitionStatsFile(
         table, snapshot.snapshotId(), schema(partitionType), sortedStats);
   }
 
+  private static PartitionMap<PartitionStats> computeStats(
+      Table table, Snapshot snapshot, StructType partitionType, boolean recompute)
+      throws IOException {
+    if (recompute) {
+      return PartitionStatsUtil.computeStats(table, null, snapshot);
+    }
+
+    PartitionStatisticsFile statisticsFile = latestStatsFile(table, snapshot.snapshotId());
+    if (statisticsFile == null) {
+      LOG.info("Previous stats not found. Computing the stats for whole table.");
+      return PartitionStatsUtil.computeStats(table, null, snapshot);
+    }
```

Comment:
Could this throw an error instead?

Reply:
why? That is to handle the case when no stats files existed before, and we need to execute a full computation.

Reply:
If I understand correctly, the user requested incremental stat compute, but with wrong parameters. In this case we could either "correct" the mistake or throw an error.

The question is, how frequent is the problem, and how easy is to detect from the user side

Reply:
what do you mean by wrong parameters? non existing snapshotId?

Reply:
throwing error now and added the testcase

Reply:
i don't agree with that design, see https://github.com/apache/iceberg/pull/12629#discussion_r2016760966